internal/plugin/builtin/testing/testhostplugin/testing.go

```diff
@@ -0,0 +1,439 @@
+package testhostplugin
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/sdk/pbs/plugin"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+// testExpectedCatalogAttributes is the expected attributes payload
+// for the attributes section for OnCreate and OnUpdate host catalog
+// requests.
+var testExpectedCatalogAttributes = map[string]interface{}{
+	"type": "catalog",
+	"foo":  "bar",
+}
+
+// testExpectedCatalogAttributesNew is the expected attributes
+// payload for the attributes section for OnUpdate and OnDelete host
+// catalog requests.
+var testExpectedCatalogAttributesNew = map[string]interface{}{
+	"type": "catalog",
+	"foo":  "baz",
+}
+
+// testExpectedSetAttributes is the expected attributes payload for
+// the attributes section for OnCreate and OnUpdate host set
+// requests.
+var testExpectedSetAttributes = map[string]interface{}{
+	"type": "set",
+	"foo":  "bar",
+}
+
+// testExpectedSetAttributesNew is the expected attributes payload
+// for the attributes section for OnUpdate and OnDelete host set
+// requests.
+var testExpectedSetAttributesNew = map[string]interface{}{
+	"type": "set",
+	"foo":  "baz",
+}
+
+// testExpectedPersisted is the expected persisted data for functions
+// that take persisted data pre-update. See
+// testExpectedPersistedNew for data that's expected to be sent
+// back on update and should be expected post-update.
+var testExpectedPersisted = map[string]interface{}{
+	"secret": "A485613C-2C28-432E-965C-7F3707E6818E",
+}
+
+// testExpectedPersistedNew is the expected persisted data
+// post-update.
+var testExpectedPersistedNew = map[string]interface{}{
+	"secret": "617DC271-0531-4BCF-9539-E84D23150AEC",
+}
+
+// mapAsStruct returns a structpb.Struct from a
+// map[string]interface{}. Errors result in a panic.
+func mapAsStruct(m map[string]interface{}) *structpb.Struct {
+	out, err := structpb.NewStruct(m)
+	if err != nil {
+		panic(err)
+	}
+
+	return out
+}
+
+// TestHostPlugin is an internal plugin used for testing the host
+// plugin system.
+type TestHostPlugin struct{}
+
+// NewClient returns a HostPluginServiceClient for TestHostPlugin.
+func NewClient() plugin.HostPluginServiceClient {
+	return &TestHostPlugin{}
+}
+
+// OnCreateCatalog implements HostPluginServiceClient for TestHostPlugin.
+func (p *TestHostPlugin) OnCreateCatalog(ctx context.Context, in *plugin.OnCreateCatalogRequest, _ ...grpc.CallOption) (*plugin.OnCreateCatalogResponse, error) {
```

Comment:
The desired behavior of this method seems like it would depend on the test that is using it.  Can we remove the validation logic from these methods and instead let the users of the TestHostPlugin provide the behavior?  If no behavior is provided it could fall back to returning an unimplemented error like [generated grpc services](UnimplementedAccountServiceServer) do.  This means tests could fake a plugin which is hanging or which is returning grpc errors, none of which is possible in this current implementation.